cmd/antrea-agent/options.go

```diff
@@ -375,3 +254,182 @@ func (o *Options) validateAntreaIPAMConfig() error {
+		// the ALLOW_NO_ENCAP_WITHOUT_ANTREA_PROXY environment variable and setting it to true. This may lead to
+		// better performance when using NoEncap if Egress NetworkPolicy enforcement is not required.
+		if !features.DefaultFeatureGate.Enabled(features.AntreaProxy) {
+			if env.GetAllowNoEncapWithoutAntreaProxy() {
+				klog.InfoS("Disabling AntreaProxy in NoEncap mode will prevent Egress NetworkPolicy rules from being enforced correctly")
+			} else {
+				return fmt.Errorf("TrafficEncapMode %s requires AntreaProxy to be enabled", o.config.TrafficEncapMode)
+			}
+		}
+		if encryptionMode != config.TrafficEncryptionModeNone {
+			return fmt.Errorf("TrafficEncryptionMode %s may only be enabled in %s mode", encryptionMode, config.TrafficEncapModeEncap)
+		}
+	}
+	if o.config.NoSNAT && !(encapMode == config.TrafficEncapModeNoEncap || encapMode == config.TrafficEncapModeNetworkPolicyOnly) {
+		return fmt.Errorf("noSNAT is only applicable to the %s mode", config.TrafficEncapModeNoEncap)
+	}
+	if encapMode == config.TrafficEncapModeNetworkPolicyOnly {
+		// In the NetworkPolicyOnly mode, Antrea will not perform SNAT
+		// (but SNAT can be done by the primary CNI).
+		o.config.NoSNAT = true
+	}
+	if err := o.validateAntreaProxyConfig(); err != nil {
+		return fmt.Errorf("proxy config is invalid: %w", err)
+	}
+	if err := o.validateFlowExporterConfig(); err != nil {
+		return fmt.Errorf("failed to validate flow exporter config: %v", err)
+	}
+	if features.DefaultFeatureGate.Enabled(features.Egress) {
+		for _, cidr := range o.config.Egress.ExceptCIDRs {
+			_, _, err := net.ParseCIDR(cidr)
+			if err != nil {
+				return fmt.Errorf("Egress Except CIDR %s is invalid", cidr)
+			}
+		}
+	}
+	if features.DefaultFeatureGate.Enabled(features.NodePortLocal) {
+		startPort, endPort, err := parsePortRange(o.config.NodePortLocal.PortRange)
+		if err != nil {
+			return fmt.Errorf("NodePortLocal portRange is not valid: %v", err)
+		}
+		o.nplStartPort = startPort
+		o.nplEndPort = endPort
+	} else if o.config.NodePortLocal.Enable {
+		klog.InfoS("The nodePortLocal.enable config option is set to true, but it will be ignored because the NodePortLocal feature gate is disabled")
+	}
+	if err := o.validateAntreaIPAMConfig(); err != nil {
+		return fmt.Errorf("failed to validate AntreaIPAM config: %v", err)
+	}
+	return nil
+}
+
+// resetVMDefaultFeatures sets the feature's default enablement status as false if it is not supported on a VM or a BM.
+func (o *Options) resetVMDefaultFeatures() error {
+	disabledFeatureMap := make(map[string]bool)
+	for f, s := range features.DefaultAntreaFeatureGates {
+		if s.Default && !features.SupportedOnVM(f) {
+			disabledFeatureMap[string(f)] = false
+		}
+	}
+	return features.DefaultMutableFeatureGate.SetFromMap(disabledFeatureMap)
+}
+
+func (o *Options) validateVMOptions() error {
+	var unsupported []string
+	for f, enabled := range o.config.FeatureGates {
```

Comment:
There can be features not enabled by feature gate (e.g. encapMode, enableNodePortLocal whose feature gate is enabled by default). Could you check any such features should be validated here too?

Reply:
Added.

Reply:
I assume you check all the options and no other unsupported options?

Reply:
Made two changes: 
1. Add unsupported options check with `EnableIPSecTunnel`, although it is deprecated, the `encryptionMode` is possibly changed to TrafficEncryptionModeIPSec if it is set as true
2. Set `encryptionMode` default value as `TrafficEncryptionModeNone` in Options.run function, and load value from the configuration only when encapMode requires supporting Encap ( the ToT code actually has a useless configuration parsing with noEncap mode ).